test/gtest/lattices.cpp

```diff
@@ -48,3 +49,38 @@ TEST(BoolLattice, Join) {
   EXPECT_FALSE(lattice.join(elem, true));
   ASSERT_TRUE(elem);
 }
+
+TEST(IntLattice, GetBottom) {
+  analysis::Int32 int32;
+  EXPECT_EQ(int32.getBottom(), (int32_t)(1ll << 31));
+
+  analysis::Int64 int64;
+  EXPECT_EQ(int64.getBottom(), (int64_t)(1ll << 63));
+
+  analysis::UInt32 uint32;
+  EXPECT_EQ(uint32.getBottom(), (uint32_t)0);
+
+  analysis::UInt64 uint64;
+  EXPECT_EQ(uint64.getBottom(), (uint32_t)0);
+}
+
+TEST(IntLattice, Compare) {
+  analysis::Int32 int32;
+  EXPECT_EQ(int32.compare(-5, 42), analysis::LESS);
+  EXPECT_EQ(int32.compare(42, -5), analysis::GREATER);
+  EXPECT_EQ(int32.compare(42, 42), analysis::EQUAL);
+}
+
+TEST(IntLattice, Join) {
+  analysis::Int32 int32;
+  int elem = 0;
+
+  EXPECT_FALSE(int32.join(elem, -10));
+  ASSERT_EQ(elem, 0);
+
+  EXPECT_FALSE(int32.join(elem, 0));
+  ASSERT_EQ(elem, 0);
+
+  EXPECT_TRUE(int32.join(elem, 100));
+  ASSERT_EQ(elem, 100);
```

Comment:
How am I supposed to understand this conceptually? That is, why does joining go "up" the integer values rather than down?

Reply:
In the literature, a lattice is a set and a "less than or equal" relation that together satisfy various properties (e.g. reflexivity, transitivity, anti-symmetry, least upper bound, greatest lower bound). Since we call the lattice relation "less than or equal," it is natural to use `<=` as the default relation for the lattice of integers.

Use cases that require `>=` instead can use `Inverted<UInt32>`, introduced in a later PR. If neither ordering is appropriate, it might be a use case for `Flat<uint32_t>`, also introduced in a later PR.

Reply:
I see, thanks. I think comments that briefly mention this and related things would be useful, to not assume the reader is completely familiar with that literature.